addons/sale_coupon/models/sale_coupon_program.py

```diff
@@ -242,22 +246,28 @@ def _filter_programs_on_products(self, order):
         i.e Buy 1 imac + get 1 ipad mini free then check 1 imac is on cart or not
         or  Buy 1 coke + get 1 coke free then check 2 cokes are on cart or not
         """
-        order_lines = order.order_line.filtered(lambda line: line.product_id) - order._get_reward_lines()
```

Comment:
Well, I'll drop this one, doesn't bring much in this specific case, even if it not optimized on large SO's :D

Reply:
i just found it funny to have basically a filtered within a filtered, indeed your solution is cleaner and probably more performant, but if it doesn't make a big difference then it might be best for a refactoring in the future (maybe)